pkg/network/server.go

```diff
@@ -757,6 +771,104 @@ func (s *Server) handleGetDataCmd(p Peer, inv *payload.Inventory) error {
 	return nil
 }
 
+// handleGetMPTDataCmd processes the received MPT inventory.
+func (s *Server) handleGetMPTDataCmd(p Peer, inv *payload.MPTInventory) error {
+	if !s.chain.GetConfig().P2PStateExchangeExtensions {
+		return errors.New("GetMPTDataCMD was received, but P2PStateExchangeExtensions are disabled")
+	}
+	resp := payload.MPTData{}
+	size := payload.MaxSize - 8 // max(io.GetVarSize(len(nodes)))
+	for _, h := range inv.Hashes {
+		if size <= 2 { // at least 1 byte for len(nodeBytes) and 1 byte for node type
+			break
+		}
+		err := s.chain.GetStateModule().Traverse(h,
+			func(node []byte) {
+				resp.Nodes = append(resp.Nodes, node)
+				l := len(node)
+				size -= l + io.GetVarSize(l)
+			}, func(node []byte) bool {
+				l := len(node)
+				return l+io.GetVarSize(l) > size
+			})
+		if err != nil {
+			return fmt.Errorf("failed to traverse MPT starting from %s: %w", h.StringBE(), err)
+		}
+	}
+	if len(resp.Nodes) > 0 {
+		msg := NewMessage(CMDMPTData, &resp)
+		return p.EnqueueP2PMessage(msg)
+	}
+	return nil
+}
+
+// handleMPTDataCmd processes the received MPT data.
+func (s *Server) handleMPTDataCmd(p Peer, data *payload.MPTData) error {
+	if !s.chain.GetConfig().P2PStateExchangeExtensions {
+		return errors.New("MPTDataCMD was received, but P2PStateExchangeExtensions are disabled")
+	}
+	toBeRequested := make(map[util.Uint256]*mptpool.Item)
+	received := make(map[util.Uint256]bool)
+	for _, nBytes := range data.Nodes {
+		var n mpt.NodeObject
+		r := io.NewBinReaderFromBuf(nBytes)
+		n.DecodeBinary(r)
+		if r.Err != nil {
+			return fmt.Errorf("failed to decode MPT node: %w", r.Err)
+		}
+		if n.Type() == mpt.HashT {
+			return errors.New("HashNode received")
+		}
+		itm, ok := s.mptPool.TryGet(n.Hash())
+		if !ok {
+			if itm, ok = toBeRequested[n.Hash()]; !ok {
+				continue
```

Comment:
It is, and `handleGetMPTData` command sends only those items that can be verified (i.e. that are a part of the requested tree). But at the same time if someone sends us a broken message with an unknown node then we can just skip it and iterate over other nodes because the message can still contain the nodes we need.

So do we still need to return an error instead?

Reply:
If someone sends us a broken message I'd rather close connection that someone, who knows what can we expect from him.